provider/datagen/src/options.rs

```diff
@@ -0,0 +1,184 @@
+        Options {
+            keys: Default::default(),
+            locales: Default::default(),
+            dictionary_segmenter: Default::default(),
+            trie_type: IcuTrieType::Small,
+            collation_han_database: CollationHanDatabase::Implicit,
+            collations: Default::default(),
+        }
+    }
+}
+
+impl SourceData {
+    // This should be on Options and take a &mut SourceData, but because of aliasing the call source.options.resolve(&source) doesn't compile.
+    pub(crate) fn resolve(&mut self) -> Result<(), DataError> {
+        self.options.locales = match core::mem::take(&mut self.options.locales) {
+            LocaleInclude::None => LocaleInclude::Explicit(Default::default()),
+            LocaleInclude::CldrSet(levels) => LocaleInclude::Explicit(
+                self.locales(levels.iter().copied().collect::<Vec<_>>().as_slice())?
+                    .into_iter()
+                    .collect(),
+            ),
+            s => s,
+        };
+        Ok(())
+    }
+}
+
+impl Options {
+    pub(crate) fn resolved_keys(&self) -> HashSet<DataKey> {
+        match &self.keys {
+            KeyInclude::None => Default::default(),
+            KeyInclude::All => crate::all_keys().into_iter().collect(),
+            KeyInclude::AllWithExperimental => {
+                crate::all_keys_with_experimental().into_iter().collect()
+            }
+            KeyInclude::Explicit(set) => set.clone(),
+        }
+    }
+}
+
+/// Defines the keys that will be generated
+#[non_exhaustive]
+#[derive(Clone, Debug, PartialEq)]
+pub enum KeyInclude {
+    /// No keys
+    None,
+    /// All stable keys
+    All,
+    /// All stable and experimental keys
+    AllWithExperimental,
+    /// An explicit set of keys.
+    Explicit(HashSet<DataKey>),
+}
+
+impl Default for KeyInclude {
+    fn default() -> Self {
+        Self::All
+    }
+}
+
+#[allow(dead_code)] // serde not in use yet
+mod data_key_as_str {
+    use super::*;
+    use serde::{de::*, ser::*};
+    use std::borrow::Cow;
+
+    pub fn serialize<S: Serializer>(selff: &HashSet<DataKey>, ser: S) -> Result<S::Ok, S::Error> {
+        selff
+            .iter()
+            .map(|k| k.path().get())
+            .collect::<HashSet<_>>()
+            .serialize(ser)
+    }
+
+    pub fn deserialize<'de, D: Deserializer<'de>>(de: D) -> Result<HashSet<DataKey>, D::Error> {
+        HashSet::<Cow<'de, str>>::deserialize(de)?
+            .into_iter()
+            .map(|s| crate::key(&s).ok_or(s))
+            .collect::<Result<_, _>>()
+            .map_err(|s| D::Error::custom(format!("Unknown key {s}")))
+    }
+}
+
+/// Defines the locaes to include
+#[non_exhaustive]
+#[derive(Clone, Debug, PartialEq)]
+pub enum LocaleInclude {
+    /// All locales
+    All,
+    /// No locales
+    None,
+    /// An explicit set of locales
+    Explicit(HashSet<LanguageIdentifier>),
+    /// All locales with the given CLDR coverage levels
+    CldrSet(HashSet<CoverageLevel>),
+}
```

Comment:
Nit: I'd still like to see this support mixed types of includes like `--locales modern ff-Adlm`.

Question: How do you see `Auto` mode fitting in here?

Reply:
`Auto` will just be another variant that we can globally resolve to `Explicit`, but I'd rather let each key decide what it suggests for `Auto` by matching on this.

Mixed types are kinda annoying on the Rust API, because you always have to wrap in another enum. We can already implement `--locales modern ff-Adlm`, and it's already possible through the Rust API by the client doing their own mixing (with `SourceData::locales(coverage_levels)`. The only thing that mixed types would add would be to have JSON like
```
  locales: [ "modern", "ff-Adlm" ]
```